include/alarm_heap.hpp

```diff
@@ -0,0 +1,198 @@
+* the program along with SSL, set forth below. This program is distributed
+* in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+* without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+* A PARTICULAR PURPOSE. See the GNU General Public License for more
+* details. You should have received a copy of the GNU General Public
+* License along with this program. If not, see
+* <http://www.gnu.org/licenses/>.
+*
+* The author can be reached by email at clearwater@metaswitch.com or by
+* post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+*
+* Special Exception
+* Metaswitch Networks Ltd grants you permission to copy, modify,
+* propagate, and distribute a work formed by combining OpenSSL with The
+* Software, or a work derivative of such a combination, even if such
+* copying, modification, propagation, or distribution would otherwise
+* violate the terms of the GPL. You must comply with the GPL in all
+* respects for all of the code used other than OpenSSL.
+* "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+* Project and licensed under the OpenSSL Licenses, or a work based on such
+* software and licensed under the OpenSSL Licenses.
+* "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+* under which the OpenSSL Project distributes the OpenSSL toolkit software,
+* as those licenses appear in the file LICENSE-OPENSSL.
+*/
+
+#ifndef ALARM_HEAP_HPP
+#define ALARM_HEAP_HPP
+
+#include <time.h>
+
+#include <string>
+#include <map>
+#include <cstdint>
+
+#include "alarm_table_defs.hpp"
+#include "alarm_trap_sender.hpp"
+#include "timer_heap.h"
+#include "utils.h"
+#include "cond_var.h"
+
+enum AlarmSeverityChange
+{
+  NO_CHANGE,
+  INCREASED_SEVERITY,
+  REDUCED_SEVERITY,
+};
+
+
+class AlarmInHeap : public HeapableTimer
+{
+public:
+  AlarmInHeap(unsigned int index) :
+    _pop_time(UINT64_MAX),
+    _severity(AlarmDef::Severity::UNDEFINED_SEVERITY),
+    _index(index)
+ {}
+
+  // Updates the pop time of the AlarmInHeap, and rebalances the heap to
+  // account for the changed pop time
+  void update_pop_time(uint64_t time_to_add)
+  {
+    _pop_time = Utils::get_time() + time_to_add;
+    _heap->rebalance(this);
+  }
+
+  // Removes the AlarmInHeap from the heap, and sets the pop time/severity
+  // to clearly invalid values (that won't cause INFORMS to be sent in error
+  // cases)
+  void remove_from_heap()
+  {
+    _pop_time = UINT64_MAX;
+    _severity = AlarmDef::Severity::UNDEFINED_SEVERITY;
+    _heap->remove(this);
+  }
+
+  const unsigned int index() { return _index; }
+  AlarmDef::Severity severity() { return _severity; }
+  void set_severity(AlarmDef::Severity severity) { _severity = severity; }
+  bool in_heap() { return (_heap != nullptr); }
+  uint64_t get_pop_time() const { return _pop_time; }
+
+private:
+  uint64_t _pop_time;
+  AlarmDef::Severity _severity;
+  unsigned int _index;
+};
+
+class AlarmInfo
```

Comment:
Actually, thinking about file layout - should we group all the classes that somehow represent a single alarm (AlarmTableDef, AlarmInHeap and AlarmInfo) into one file, and all the classes that are alarm containers (AlarmTableDefs, AlarmHeap) into another?